llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,386 @@
+  /// register (i.e. $x0). This can cause issues, when detecting Write
+  /// After Write (WAW) dependencies. To mitigate this, only use the last
+  /// definition of a virtual register to count the definitions, so that
+  /// writing into subregisters does not already trigger the register
+  /// replacement. In this case $x0 would have already been replaced, even
+  /// though there is no real WAW dependency.
+  ///                                                                          \
+  /// undef %0.sub_256_lo:mxa, _, _= VLDA_2D_dmw_lda_w _,_                     \
+  /// % 0.sub_256_hi:vec512 = VLDA_dmw_lda_w_ag_idx_imm _,_                    \
+  /// %7:mxm, %8:el = VMAX_LT_D8 %0, $x6                                       \
+  ///                                                                          \
+  /// returns a mapping between a virtual register and its last defining machine
+  /// instruction.
+  std::map<Register, const MachineInstr *>
+  getLastVRegDef(const MachineBasicBlock &MBB) const;
+
+  /// Block every sub- and super-register of a physical register, so that it is
+  /// removed for future replacement strategies, i.e. block wl4, wh4, y2 if X4
+  /// is used.
+  void addAliasRegs(BitVector &BlockedPhysRegs, const MCPhysReg PhysReg) const;
+};
+
+MCPhysReg
+AIEWawRegRewriter::getAssignedPhysReg(const MachineOperand &MO) const {
+  assert(MO.isReg());
+  Register Reg = MO.getReg();
+
+  if (Reg.isStack())
+    return MCRegister::NoRegister;
+
+  if (Reg.isVirtual())
+    Reg = VRM->getPhys(Reg);
+
+  return Reg;
+}
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  SmallVector<const MachineBasicBlock *, 4> LoopMBBs =
+      AIELoopUtils::getSingleBlockLoopMBBs(MF);
+
+  if (LoopMBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+  bool Modified = false;
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopMBBs)
+    Modified |= renameMBBPhysRegs(MBB);
+
+  return Modified;
+}
+
+/// return a vector containing virtual registers, that have had a copy to define
+/// the virtual register
+SmallVector<Register, 16> getVRegWithCopies(const MachineBasicBlock &MBB) {
+  SmallVector<Register, 16> VRegWithCopies;
+  for (const MachineInstr &MI : MBB) {
+    for (const MachineOperand Def : MI.defs()) {
+      Register Reg = Def.getReg();
+
+      if (!Reg.isVirtual())
+        continue;
+
+      if (MI.isCopyLike() && find(VRegWithCopies, Reg) == VRegWithCopies.end())
+        VRegWithCopies.push_back(Reg);
+    }
+  }
+
+  return VRegWithCopies;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
```

Comment:
The heuristic I see here is "if a register is used in a copy don't change it".
Spoke with Fabian offline about it.
The current heuristic would still rename a register used in a copy if there are several inner loops in the kernel.

At this point, we assume this is not something we care about.

One quick fix could be to keep track of the copies for all the single loop BBs instead of computing the list of copies on the fly for each MBB.